keydata_test.go

```diff
@@ -1427,3 +1562,276 @@ func (s *keyDataSuite) TestReadAndWriteWithLegacySnapModelAuthKey(c *C) {
+			`"authorized_snap_models":{`+
+			`"alg":"sha256",`+
+			`"kdf_alg":"sha256",`+
+			`"key_digest":{`+
+			`"alg":"sha256",`+
+			`"salt":"IPDKKUOoRYwvMWX8LoCCtlGgzgzokAhsh42XnbGUn0s=",`+
+			`"digest":"SSbv/yS8h5pqchVfV9AMHUjhS/vVateojNRRmo624qk="},`+
+			`"hmacs":["JWziaukXiAIsPU22X1RTC/2wEkPN4IdNvgDEzSnWXIc="]}}
+`))
+}
+
+func (s *keyDataSuite) TestMakeDiskUnlockKey(c *C) {
+	primaryKey := testutil.DecodeHexString(c, "1850fbecbe8b3db83a894cb975756c8b69086040f097b03bd4f3b1a3e19c4b86")
+	kdfAlg := crypto.SHA256
+	unique := testutil.DecodeHexString(c, "1850fbecbe8b3db83a894cb975756c8b69086040f097b03bd4f3b1a3e19c4b86")
+
+	reader := new(bytes.Buffer)
+	reader.Write(unique)
+
+	unlockKey, clearTextPayload, err := MakeDiskUnlockKey(reader, kdfAlg, primaryKey)
+	c.Assert(err, IsNil)
+
+	knownGoodUnlockKey := testutil.DecodeHexString(c, "8b78ddabd8e38a6513e654638c0f7b8c738d5461a403564d19d98e7f8ed469cb")
+	c.Check(unlockKey, DeepEquals, DiskUnlockKey(knownGoodUnlockKey))
+
+	knownGoodPayload := testutil.DecodeHexString(c, "304404201850fbecbe8b3db83a894cb975756c8b69086040f097b03bd4f3b1a3e19c4b8604201850fbecbe8b3db83a894cb975756c8b69086040f097b03bd4f3b1a3e19c4b86")
+	c.Check(clearTextPayload, DeepEquals, knownGoodPayload)
+
+	st := cryptobyte.String(clearTextPayload)
+	c.Assert(st.ReadASN1(&st, cryptobyte_asn1.SEQUENCE), Equals, true)
+
+	var p PrimaryKey
+	c.Assert(st.ReadASN1Bytes((*[]byte)(&p), cryptobyte_asn1.OCTET_STRING), Equals, true)
+	c.Check(p, DeepEquals, PrimaryKey(primaryKey))
+
+	var u []byte
+	c.Assert(st.ReadASN1Bytes(&u, cryptobyte_asn1.OCTET_STRING), Equals, true)
+	c.Check(u, DeepEquals, unique)
+}
+
+type keyDataLegacySuite struct {
+	keyDataSuite
+}
+
+func (s *keyDataLegacySuite) SetUpSuite(c *C) {
+	s.keyDataSuite.SetUpSuite(c)
+	s.Version = 1
+}
+
+func (s *keyDataLegacySuite) SetUpTest(c *C) {
+	s.AddCleanup(MockReadKeyData(s.keyDataTestBase.Version))
+}
+
+var _ = Suite(&keyDataLegacySuite{})
+
+type testLegacyKeyPayloadData struct {
+	key    DiskUnlockKey
+	auxKey PrimaryKey
+}
+
+func (s *keyDataLegacySuite) testKeyPayload(c *C, data *testKeyPayloadData) {
+	unlockKey := data.unique
+	primaryKey := data.primary
+
+	payload := MarshalV1Keys(unlockKey, primaryKey)
+
+	key, auxKey, err := UnmarshalV1KeyPayload(payload)
+	c.Check(err, IsNil)
+	c.Check(key, DeepEquals, unlockKey)
+	c.Check(auxKey, DeepEquals, primaryKey)
+}
+
+func (s *keyDataLegacySuite) TestRecoverKeysWithPassphraseAuthModeNone(c *C) {
+	auxKey := s.newPrimaryKey(c, 32)
+	protected, _ := s.mockProtectKeys(c, auxKey)
+
+	keyData, err := NewKeyData(protected)
+	c.Assert(err, IsNil)
+	recoveredKey, recoveredAuxKey, err := keyData.RecoverKeysWithPassphrase("", nil)
+	c.Check(err, ErrorMatches, "cannot recover key with passphrase")
+	c.Check(recoveredKey, IsNil)
+	c.Check(recoveredAuxKey, IsNil)
+}
+
+type keyDataScopeSuite struct {
+	keyDataSuite
+}
+
+func (s *keyDataScopeSuite) SetUpTest(c *C) {
```

Comment:
As mentioned elsewhere, these kinds of tests should be in the bootenv package.

Reply:
I moved all of them to bootenv. I remember talking about having bootenv potentially as a different package at some point so I copied all the bits I needed from `keydata_test.go` as is. Let me know what you think or if we should create some common package to increase reuse.